File: frontend/src/types/session.ts
export type SessionStepStatus = 'queued' | 'running' | 'success' | 'failed' | string;

export interface SessionCreate {
  title?: string | null;
  description?: string | null;
}

export interface SessionStepRead {
  id: string; // UUID
  order: number;
  tool: string;
  step: string;
  substep?: string | null;
  algorithm?: string | null;
  params?: Record<string, any> | null;
  status: SessionStepStatus;
  started_at?: string | null;
  finished_at?: string | null;
  error?: string | null;
  run_ref_type?: string | null;
  run_ref_id?: string | null;
}

export interface SessionRead {
  id: string; // UUID
  title?: string | null;
  description?: string | null;
  created_at?: string | null;
  closed_at?: string | null;
  steps: SessionStepRead[];
}

export interface SaveWorkflowFromSessionRequest {
  name: string;
  description?: string | null;
  selected_step_ids?: string[]; // if omitted, include all steps
}

File: backend/app/schemas/session.py
from typing import Any, Dict, List, Optional
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


class SessionCreate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None

class SessionStepCreate(BaseModel):
    tool: str
    step: str
    substep: Optional[str] = None
    algorithm: Optional[str] = None
    params: Optional[Dict[str, Any]] = None


class SessionStepRead(BaseModel):
    id: UUID
    order: int
    tool: str
    step: str
    substep: Optional[str] = None
    algorithm: Optional[str] = None
    params: Optional[Dict[str, Any]] = None
    status: str
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    error: Optional[str] = None
    run_ref_type: Optional[str] = None
    run_ref_id: Optional[str] = None

    # Pydantic v2: allow serialization from ORM objects
    model_config = ConfigDict(from_attributes=True)


class SessionRead(BaseModel):
    id: UUID
    title: Optional[str] = None
    description: Optional[str] = None
    created_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    steps: List[SessionStepRead] = Field(default_factory=list)

    # Pydantic v2: allow serialization from ORM objects
    model_config = ConfigDict(from_attributes=True)


class SaveWorkflowFromSessionRequest(BaseModel):
    name: str
    description: Optional[str] = None
    selected_step_ids: Optional[List[UUID]] = None  # if None, include all; else include only these
